### PR TITLE
Fixing poorly uncommented code from previous MR

### PR DESCRIPTION
**Pull request by thomasgudjonwright**
This fixes a bug from the previously merged PR that added compats. This addresses CI issues as well

### CODE REVIEW

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -71,13 +71,14 @@ jobs:
...
-          version: '1'
+          version: '1.6'
       - run: |
           julia --project=docs -e '
             using Pkg
             Pkg.develop(PackageSpec(path=pwd()))
             Pkg.instantiate()
             include("docs/make.jl")'
         env:
+          JULIA_PKG_SERVER: ""
```
Comment (mzgubic) [inline .github/workflows/CI.yml:82]: I know nothing about CI, what does this do?

File: src/ilmm.jl [hunk 2]
```diff
@@ -29,7 +29,10 @@ julia> f = ILMM(IndependentMOGP([GP(SEKernel())]), rand(2,2));
 
 julia> latent_f = get_latent_gp(f);
 
-julia> latent_f == IndependentMOGP([GP(SEKernel())])
+julia> latent_f isa IndependentMOGP
```
Comment (mzgubic) [inline src/ilmm.jl:32]: why do we want this change? I find the previous one clearer personally

**Review by mzgubic — APPROVED** [review]

File: src/ilmm.jl (hunk 2 above)
Comment (thomasgudjonwright) [reply src/ilmm.jl:32]: Something was going on with equality comparisons of the composite types I introduced, so I changed the doctests to make sure they were equivalent instead of "equal"

Comment (mzgubic) [reply src/ilmm.jl:32]: hmm, what exactly was going on? is there an issue about this? we might want to fix that instead?

File: .github/workflows/CI.yml (hunk 1 above)
Comment (thomasgudjonwright) [reply .github/workflows/CI.yml:82]: See [this thread](https://julialang.slack.com/archives/C67TK21LJ/p1629306333256900) for proper answer to this question. 

TLDR: The package server can be out of date, meaning that the CI (depending on how its built) can miss recent releases, introducing compat issues that don't actually exist

File: src/ilmm.jl (hunk 2 above)
Comment (thomasgudjonwright) [reply src/ilmm.jl:32]: Given this
```
struct IndependentMOGP{Tfs<:Vector{<:AbstractGP}} <: AbstractGP
    fs::Tfs
end
function independent_mogp(fs::Vector{<:AbstractGP})
    return IndependentMOGP(fs)
end
```
Should we expect:
```
julia> independent_mogp([GP(KernelFunctions.SEKernel())]) == IndependentMOGP([GP(KernelFunctions.SEKernel())])
true
```
I guess yes, and I am mistaking `isequal`, `==` and `===`

Comment (thomasgudjonwright) [reply src/ilmm.jl:32]: This is what made me think composite types couldn't be compared directly.
```
julia> independent_mogp([GP(SEKernel())])
IndependentMOGP{Vector{GP{AbstractGPs.ZeroMean{Float64}, SqExponentialKernel{Distances.Euclidean}}}}(GP{AbstractGPs.ZeroMean{Float64}, SqExponentialKernel{Distances.Euclidean}}[GP{AbstractGPs.ZeroMean{Float64}, SqExponentialKernel{Distances.Euclidean}}(AbstractGPs.ZeroMean{Float64}(), Squared Exponential Kernel (metric = Distances.Euclidean(0.0)))])

julia> IndependentMOGP([GP(SEKernel())])
IndependentMOGP{Vector{GP{AbstractGPs.ZeroMean{Float64}, SqExponentialKernel{Distances.Euclidean}}}}(GP{AbstractGPs.ZeroMean{Float64}, SqExponentialKernel{Distances.Euclidean}}[GP{AbstractGPs.ZeroMean{Float64}, SqExponentialKernel{Distances.Euclidean}}(AbstractGPs.ZeroMean{Float64}(), Squared Exponential Kernel (metric = Distances.Euclidean(0.0)))])

julia> independent_mogp([GP(SEKernel())]) == IndependentMOGP([GP(SEKernel())])
false

julia> isequal(independent_mogp([GP(SEKernel())]), IndependentMOGP([GP(SEKernel())]))
false

julia> independent_mogp([GP(SEKernel())]) === IndependentMOGP([GP(SEKernel())])
false
```

Comment (mzgubic) [reply src/ilmm.jl:32]: Yes, I would expect that. `==` checks for equality of two things (and defaults to `===`), while `===` checks that two things are aliased (i.e. the same object in memory).
So I would expect
```julia
julia> independent_mogp([GP(SEKernel())]) == IndependentMOGP([GP(SEKernel())])
true

julia> independent_mogp([GP(SEKernel())]) === IndependentMOGP([GP(SEKernel())])
false
```

Comment (mzgubic) [reply src/ilmm.jl:32]: In case you are getting
```julia
julia> independent_mogp([GP(SEKernel())]) == IndependentMOGP([GP(SEKernel())])
false
```
we should probably define an equality `==` to compare the values rather than check for `===`, so that this doesn't trip us in other places

Comment (thomasgudjonwright) [reply src/ilmm.jl:32]: I feel like we shouldn't have to define our own `==` and that it should work for composite types by default. Maybe I should ask in #julia why this might not be working

Comment (mzgubic) [reply src/ilmm.jl:32]: oh sorry didn't see your second reply - we should define `Base.:(==)(a::MOGP, b::MOGP)`

Comment (mzgubic) [reply src/ilmm.jl:32]: It does not work by default for composite types unfortunately.
see e.g.
```julia
julia> struct MyStruct
           a
       end

julia> a = MyStruct(1.0)
MyStruct(1.0)

julia> b = MyStruct(1.0)
MyStruct(1.0)

julia> @which ==(a, b)
==(x, y) in Base at Base.jl:87
```
which is the default that you can inspect with
```julia
@edit ==(a, b)

# shows
# ==(x, y) = x === y
```

Comment (thomasgudjonwright) [reply src/ilmm.jl:32]: So you think its worth adding this by comparing fields and type rather than just explicitly checking in the tests? I definitely can do so, I just feel like I haven't seen an `==` dispatch defined for most composite types I have come across.

File: .github/workflows/CI.yml [hunk 3]
```diff
@@ -71,13 +71,14 @@ jobs:
       - uses: actions/checkout@v2
       - uses: julia-actions/setup-julia@v1
         with:
-          version: '1'
+          version: '1.6'
```
Comment (willtebbutt) [inline .github/workflows/CI.yml:74]: I think this should probably be `1`, as opposed to `1.6`.

`1` means "the latest version in the 1.x series of releases", so at the moment it will give `1.6`. When `1.7` is released, it'll do that.

Comment (willtebbutt) [reply .github/workflows/CI.yml:74]: Actually, this should probably just be `1` and `1.6`, so that we don't accidentally drop support for `1.6` before we want to.

File: src/ilmm.jl [hunk 4]
```diff
@@ -43,7 +46,7 @@ Return the diagonal element of Σ.
 ```jldoctest
 julia> Σ = Diagonal(Fill(2, 3));
 
-julia> noise_var(Σ) == 2
+julia> LinearMixingModels.noise_var(Σ) == 2
```
Comment (willtebbutt) [inline src/ilmm.jl:49]: My advice would be against using internal details in the docstrings. In general, stuff that users shouldn't have to know about probably shouldn't be in the docstrings. So things like this and `reshape_y` should probably just be tested in the tests.

**Review by willtebbutt — COMMENTED** [review]
Just a couple of small things.

File: src/ilmm.jl (hunk 2 above)
Comment (mzgubic) [reply src/ilmm.jl:32]: I don't know, it seems pretty common to me (but maybe we have been working in different packages where conventions differ). e.g. [ChainRulesCore defines `==` for `Tangent` type](https://github.com/JuliaDiff/ChainRulesCore.jl/blob/master/src/differentials/composite.jl#L48-L55), so does [NamedDims](https://github.com/invenia/NamedDims.jl/blob/master/src/functions.jl#L87-L94) for it's wrapper array.

Maybe it's less common in higher level packages. The only downside I can really think of is invalidations, but I don't expect it to be a serious issue.

Comment (thomasgudjonwright) [reply src/ilmm.jl:32]: @willtebbutt do you have any thoughts on this? Should equality be defined for an `IndependentMOGP`? If so then also for `Orthogonals`, and `ILMMs`?

Comment (willtebbutt) [reply src/ilmm.jl:32]: Yeah -- I don't think one should avoid defining checks for equality, but it's pretty reasonable to ask whether / why you need it. In the wild, are you ever likely to need to check for equality between two IndependentMOGPs, or is this just for testing purposes? If it's the latter, I wouldn't bother to write the code as it's just more code.

Comment (thomasgudjonwright) [reply src/ilmm.jl:32]: I can't really see why anyone would want to check between the two, but I could certainly be missing cases of when one might. It was really just for testing a super simple one-line constructor helper function!

Comment (willtebbutt) [reply src/ilmm.jl:32]: Tbh, we've not defined the equality operator in AbstractGPs, and it has yet to cause any problems. I would be inclined not to bother, and to test this constuctor in a different manner 🤷

File: .github/workflows/CI.yml (hunk 3 above)
Comment (thomasgudjonwright) [reply .github/workflows/CI.yml:74]: So you're thinking:
```
version: 
   - '1'
   - '1.6'
``` 
?

Comment (willtebbutt) [reply .github/workflows/CI.yml:74]: Exactly.

File: src/ilmm.jl (hunk 4 above)
Comment (thomasgudjonwright) [reply src/ilmm.jl]: Done

File: .github/workflows/CI.yml (hunk 3 above)
Comment (thomasgudjonwright) [reply .github/workflows/CI.yml]: Done

File: src/ilmm.jl (hunk 2 above)
Comment (thomasgudjonwright) [reply src/ilmm.jl:32]: Resolving for now then.

File: .github/workflows/CI.yml (hunk 3 above)
Comment (thomasgudjonwright) [reply .github/workflows/CI.yml]: This seems to have broken the `CI`. Adding the second version must be more than that one liner.

File: .github/workflows/CI.yml [hunk 5]
```diff
@@ -71,13 +71,16 @@ jobs:
       - uses: actions/checkout@v2
       - uses: julia-actions/setup-julia@v1
         with:
-          version: '1'
+          version:
+            - '1'
+            - '1.6'
```
Comment (willtebbutt) [inline .github/workflows/CI.yml:76]:
```suggestion
            - '1'
```
Sorry, I thought this was part of the main testing bit rather than the docs 🤦 . Version `1` is fine for building the docs.

File: .github/workflows/CI.yml (hunk 3 above)
Comment (thomasgudjonwright) [reply .github/workflows/CI.yml]: Resolving since discussed in a thread below

File: .github/workflows/CI.yml (hunk 5 above)
Comment (thomasgudjonwright) [reply .github/workflows/CI.yml]: Fixed, resolving

**Review by willtebbutt — APPROVED** [review]
LGTM. Let's get it registered.
